File: backend/src/app/routers/logs.py
from datetime import date
from uuid import UUID

from app.dependencies.auth import get_current_user_id
from app.dependencies.supabase_client import get_authenticated_supabase
from app.repositories import log_repo
from app.schemas.log import (
    CreateMealLogRequest,
    CreateWorkoutLogRequest,
    MealLogResponse,
    WorkoutLogResponse,
)
from fastapi import APIRouter, Depends, Query, status

from supabase import AsyncClient

router = APIRouter(prefix="/logs", tags=["logs"])


@router.post("/meal", status_code=status.HTTP_201_CREATED, response_model=MealLogResponse)
async def create_meal_log(
    body: CreateMealLogRequest,
    user_id: UUID = Depends(get_current_user_id),
    supabase: AsyncClient = Depends(get_authenticated_supabase),
) -> MealLogResponse:
    return await log_repo.upsert_meal_log(supabase, user_id, body)


@router.get("/meal", response_model=dict)
async def get_meal_logs(
    log_date: date = Query(...),
    user_id: UUID = Depends(get_current_user_id),
    supabase: AsyncClient = Depends(get_authenticated_supabase),
) -> dict:
    logs = await log_repo.get_meal_logs_by_date(supabase, user_id, log_date)
    return {"logs": logs}


@router.post("/workout", status_code=status.HTTP_201_CREATED, response_model=WorkoutLogResponse)
async def create_workout_log(
    body: CreateWorkoutLogRequest,
    user_id: UUID = Depends(get_current_user_id),
    supabase: AsyncClient = Depends(get_authenticated_supabase),
) -> WorkoutLogResponse:
    return await log_repo.upsert_workout_log(supabase, user_id, body)


@router.get("/workout", response_model=dict)
async def get_workout_logs(
    log_date: date = Query(...),
    user_id: UUID = Depends(get_current_user_id),
    supabase: AsyncClient = Depends(get_authenticated_supabase),
) -> dict:
    logs = await log_repo.get_workout_logs_by_date(supabase, user_id, log_date)
    return {"logs": logs}

File: frontend/src/lib/api/logs.ts
import { apiClient } from './client'
import type {
  AdaptationResponse,
  CreateMealLogRequest,
  CreateFeedbackRequest,
  CreateWorkoutLogRequest,
  FeedbackEventDetailResponse,
  FeedbackTagResponse,
  MealLogResponse,
  WorkoutLogResponse,
} from '@/types/log'

export async function createMealLog(
  token: string,
  data: CreateMealLogRequest
): Promise<MealLogResponse> {
  return apiClient<MealLogResponse>('/logs/meal', {
    method: 'POST',
    body: JSON.stringify(data),
  }, token)
}

export async function getMealLogs(
  token: string,
  logDate: string
): Promise<{ logs: MealLogResponse[] }> {
  return apiClient<{ logs: MealLogResponse[] }>(
    `/logs/meal?log_date=${logDate}`,
    {},
    token
  )
}

export async function createWorkoutLog(
  token: string,
  data: CreateWorkoutLogRequest
): Promise<WorkoutLogResponse> {
  return apiClient<WorkoutLogResponse>('/logs/workout', {
    method: 'POST',
    body: JSON.stringify(data),
  }, token)
}

export async function getWorkoutLogs(
  token: string,
  logDate: string
): Promise<{ logs: WorkoutLogResponse[] }> {
  return apiClient<{ logs: WorkoutLogResponse[] }>(
    `/logs/workout?log_date=${logDate}`,
    {},
    token
  )
}

export async function submitFeedback(
  token: string,
  data: CreateFeedbackRequest
): Promise<AdaptationResponse> {
  return apiClient<AdaptationResponse>('/feedback', {
    method: 'POST',
    body: JSON.stringify(data),
  }, token)
}

export async function getFeedbackTags(
  token: string,
  planId: string
): Promise<{ tags: FeedbackTagResponse[] }> {
  return apiClient<{ tags: FeedbackTagResponse[] }>(
    `/feedback/${planId}`,
    {},
    token
  )
}

export async function getFeedbackHistory(
  token: string,
  limit = 20
): Promise<FeedbackEventDetailResponse[]> {
  return apiClient<FeedbackEventDetailResponse[]>(
    `/feedback/history?limit=${limit}`,
    {},
    token
  )
}

export async function getFeedbackHistoryDetail(
  token: string,
  eventId: string
): Promise<FeedbackEventDetailResponse> {
  return apiClient<FeedbackEventDetailResponse>(
    `/feedback/history/${eventId}`,
    {},
    token
  )
}
